### ISSUE TITLE
java.lang.IllegalStateException: Unable to get number of of rows for a non 2d matrix

### ISSUE THREAD

**Issue by Zaprogramowany27** [deeplearning4j/deeplearning4j#2010]
I have neural network like follows

     public static void main(String[] args) throws IOException, InterruptedException {
        int batchSize = 19;
        int epoch = 10;
        long seed = 123;
        int height = 32;
        int width = 65;
        int channels = 3;
        int outputNumber = 4;

        Random randNumGen = new Random(seed);

        File parentDir = new File("images");
        FileSplit filesInDir = new FileSplit(parentDir, new String[]{"png"}, randNumGen);
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();
        BalancedPathFilter pathFilter = new BalancedPathFilter(randNumGen, new String[]{"png"}, labelMaker);

        InputSplit[] filesInDirSplit = filesInDir.sample(pathFilter, 80, 20);
        InputSplit trainData = filesInDirSplit[0];
        InputSplit testData = filesInDirSplit[1];

        ImageRecordReader recordReader = new ImageRecordReader(height, width, channels, labelMaker);
        recordReader.initialize(trainData);

        DataSetIterator iterator = new RecordReaderDataSetIterator(recordReader, batchSize, 0, outputNumber);

        MultiLayerConfiguration config = new NeuralNetConfiguration.Builder()
            .seed(seed)
            .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
            .iterations(1)
            .learningRate(0.006)
            .updater(Updater.NESTEROVS).momentum(0.9)
            .regularization(true).l2(1e-4)
            .list()
            .layer(0, new DenseLayer.Builder()
                .nIn(width * height)
                .nOut(10000)
                .activation("relu")
                .weightInit(WeightInit.XAVIER)
                .build()
            )
            .layer(1, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                .nIn(10000)
                .nOut(outputNumber)
                .activation("softmax")
                .weightInit(WeightInit.XAVIER)
                .build()
            )
            .pretrain(false).backprop(true)
            .build();

        MultiLayerNetwork model = new MultiLayerNetwork(config);
        model.init();
        model.setListeners(new ScoreIterationListener(1));

        for (int i = 0; i < epoch; i++) {
            model.fit(iterator);
        }
        recordReader.reset();
        recordReader.initialize(testData);

        iterator = new RecordReaderDataSetIterator(recordReader, batchSize, 0, outputNumber);
        Evaluation eval = new Evaluation(outputNumber);
        while (iterator.hasNext()) {
            DataSet next = iterator.next();
            INDArray output = model.output(next.getFeatureMatrix());
            eval.eval(next.getLabels(), output);
        }

        System.out.println(eval.stats());
    }

And directory structure:

    images:
        h: [0...1000]
        a: [0...1000]
        c: [0...1000]
        d: [0...1000]

And I got error:

    Exception in thread "main" java.lang.IllegalStateException: Unable to get number of of rows for a non 2d matrix
	at org.nd4j.linalg.api.ndarray.BaseNDArray.rows(BaseNDArray.java:3591)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.mmul(BaseNDArray.java:2467)
	at org.deeplearning4j.nn.layers.BaseLayer.preOutput(BaseLayer.java:355)
	at org.deeplearning4j.nn.layers.BaseLayer.activate(BaseLayer.java:366)
	at org.deeplearning4j.nn.layers.BaseLayer.activate(BaseLayer.java:386)

What I'm doing wrong? Is it bug?

**Comment by eraly** [comment]
@Zaprogramowany27 So this is not a bug. 
I am assuming you are reading in jpeg/png etc with the image reader. After all that jazz with the image reader is done you get your iterator. The dataset the iterator produces is off rank 4 as "batch_size*num_channels*height*width". So this in other words is a 4d matrix.
Usually images are dealt with using convolutional neural nets like the mnist convolution ex in the examples repo.
The dense layer network expects an input matrix of rank 2 - "batch_size*num_features".

So it is this mismatch in shapes that is causing your error. You have two things you can do.

1. Use conv nets to process your images since they are the state of the art anyway.
2. Write your own code to reshape your 4d matrix from your image reader into the right 2d shape. This will be a bit of a headache and you might want to make sure it is worth the headache. 
3. We have something known as a [computation graph. ](http://deeplearning4j.org/compgraph) which is meant for building a neural net that hooks up different kinds of layers, like a conv net to a dense layer etc. And you could in theory use that here just so that you don't have to write your own code.

And last thoughts are - if this is something that you think should be added as a feature that makes it more straightforward please file another issue as an enhancement outlining the use cases.

**Comment by eraly** [comment]
Closing.
